File: trigger/codegen/lib/processAudio/processAudio.c
/*
 * Academic License - for use in teaching, academic research, and meeting
 * course requirements at degree granting institutions only.  Not for
 * government, commercial, or other organizational use.
 *
 * processAudio.c
 *
 * Code generation for function 'processAudio'
 *
 */

/* Include files */
#include "processAudio.h"
#include "fft.h"
#include "filter.h"
#include "fir1.h"
#include "processAudio_data.h"
#include "processAudio_initialize.h"
#include "processAudio_rtwutil.h"
#include "processAudio_types.h"
#include "rt_nonfinite.h"
#include <math.h>

/* Function Definitions */
void processAudio(const double audio_data[44100], double Fs,
                  struct0_T *originalAudio, struct0_T *filteredAudio,
                  double *Energy)
{
  static creal_T filtered_fft_result[65536];
  double dv1[11];
  double dv[2];
  double delta1;
  int k;
  if (!isInitialized_processAudio) {
    processAudio_initialize();
  }
  /*  Calculate the length of the audio data in seconds */
  delta1 = 44100.0 / Fs;
  /*  Create a time series for plotting */
  originalAudio->Time[44099] = delta1;
  originalAudio->Time[0] = 0.0;
  if (-delta1 == 0.0) {
    delta1 /= 44099.0;
    for (k = 0; k < 44098; k++) {
      originalAudio->Time[k + 1] = (2.0 * ((double)k + 2.0) - 44101.0) * delta1;
    }
  } else if ((delta1 < 0.0) && (fabs(delta1) > 8.9884656743115785E+307)) {
    delta1 /= 44099.0;
    for (k = 0; k < 44098; k++) {
      originalAudio->Time[k + 1] = delta1 * ((double)k + 1.0);
    }
  } else {
    delta1 /= 44099.0;
    for (k = 0; k < 44098; k++) {
      originalAudio->Time[k + 1] = ((double)k + 1.0) * delta1;
    }
  }
  /*  FFT Analysis */
  /*  Number of FFT points */
  /*  Keep only the first half */
  /*  Frequency axis */
  /*  Designing Bandpass Filter */
  /*  Center frequency of the bandpass filter */
  /*  Normalize center frequency */
  delta1 = Fs / 2.0;
  dv[0] = 295.0 / delta1;
  dv[1] = 305.0 / delta1;
  /*  Bandpass filter */
  /*  Applying the Bandpass Filter */
  fir1(dv, dv1);
  filter(dv1, audio_data, filteredAudio->Amplitude);
  /*  FFT of the filtered signal */
  fft(filteredAudio->Amplitude, filtered_fft_result);
  /*  First half */
  /*  Amplitude in dB */
  /*  Calculate the energy of the filtered audio signal */
  /*  Internal function for energy calculation */
  *Energy = 0.0;
  for (k = 0; k < 32768; k++) {
    delta1 = 20.0 * log10(rt_hypotd_snf(filtered_fft_result[k].re,
                                        filtered_fft_result[k].im));
    if (delta1 > 0.0) {
      *Energy += delta1 * delta1;
      /*  Energy calculation */
    }
  }
  /*  Output structure for original and filtered audio */
  for (k = 0; k < 44100; k++) {
    originalAudio->Amplitude[k] = audio_data[k];
    filteredAudio->Time[k] = originalAudio->Time[k];
  }
}

/* End of code generation (processAudio.c) */

File: trigger/processAudio.m
function [originalAudio, filteredAudio, Energy] = processAudio(audio_data, Fs)
    % Calculate the length of the audio data in seconds
    len_time = length(audio_data) / Fs;

    % Create a time series for plotting
    time = linspace(0, len_time, length(audio_data));

    % FFT Analysis
    nfft = 2^nextpow2(length(audio_data)); % Number of FFT points
    fft_result = fft(audio_data, nfft);
    fft_result = fft_result(1:nfft/2); % Keep only the first half
    frequencies = (0:nfft/2-1) * Fs / nfft; % Frequency axis

    % Designing Bandpass Filter
    center_freq = 300; % Center frequency of the bandpass filter
    normalized_center_freq = center_freq / (Fs/2); % Normalize center frequency
    bandpass_filter = fir1(10, [295, 305] / (Fs/2), 'bandpass'); % Bandpass filter

    % Applying the Bandpass Filter
    filtered_audio = filter(bandpass_filter, 1, audio_data);

    % FFT of the filtered signal
    filtered_fft_result = fft(filtered_audio, nfft);
    filtered_fft_result = filtered_fft_result(1:nfft/2); % First half
    amplitude_filtered = 20*log10(abs(filtered_fft_result)); % Amplitude in dB

    % Calculate the energy of the filtered audio signal
    Energy = eng(1, length(amplitude_filtered), amplitude_filtered);

    % Output structure for original and filtered audio
    originalAudio = struct('Time', time, 'Amplitude', audio_data);
    filteredAudio = struct('Time', time, 'Amplitude', filtered_audio);

    % Internal function for energy calculation
    function [energ] = eng(timeLeft, timeRight, amp)
        energ = 0;
        for i = timeLeft:timeRight
            if(amp(i) > 0)
                energ = energ + amp(i) * amp(i); % Energy calculation
            end
        end
    end
end

